🛂 Update Handling of Gem Permissions

landongrindheim opened the pull request:
Until now, we've allowed a `Rubygem` to be owned by users (via `Ownership`s) and an `Organization`. This has some downsides. For example, we've had to explicitly check both ownerships and organization membership in our gem-related policies.

I've opted to push this logic into `Rubygem` and enforce the ownership check to be organization-specific if the gem belongs to an organization OR ownership-specific (i.e. user-specific) if it is not.

landongrindheim committed:
Make ownership an organization OR user concern

Until now, we've allowed a `Rubygem` to be owned by users (via
`Ownership`s) and an `Organization`. This has some downsides. For
example, we've had to explicitly check both ownerships and organization
membership in our gem-related policies.

I've opted to push this logic into `Rubygem` and enforce the ownership
check to be organization-specific if the gem belongs to an organization
OR ownership-specific (i.e. user-specific) if it is not.

pboling commented:
A downside of allowing Orgs to own gems is they have a host of distinct issues related to their own state changes that differ from the state changes that occur with people.

Keeping it as just people may be simpler than coming up with policy for how to handle a gem owned by a company that:

* goes out of business
* merges, or is acquired by, another company
* conflicting international business laws and which country's would take precedent
* lawsuits over IP, or publishing rights, that a company claims to "own"

When ownership is limited to people a policy of "for the good of the community and ecosystem" can be more easily determined and adjudicated by the rubygems team.

Orgs may complicate things. Offloading that complexity onto orgs as is currently done, may do less harm than allowing them to own gems.

Due to the additional work involved in managing Org owned gems, it might be a way for RubyGems to generate sponsorships.

Org ownership could be constrained by payment of a monthly fee, which would help alleviate the additional support/maintenance/hosting costs associated with expanding ownership to allow for orgs.

Org ownership of gems could fall back to person ownership upon non-payment of the fee.

IMO, in case it wasn't obvious, for all of the above.

landongrindheim committed:
Add role-based methods to Organization

landongrindheim committed:
Introduce GemPermissions for finer grained access

There's still work to be done, but this feels (to me) like an
improvement over our access-means-ownership model.

landongrindheim committed:
Use GemPermissions for policies

landongrindheim committed:
Allow anyone with push access to view events

landongrindheim committed:
Rename can_admin? -> can_perform_gem_admin?

This variant feels cleaner to me. Hopefully others will agree.

landongrindheim committed:
Update ownership-based policies to use permissions

landongrindheim committed:
Move to permissions-based access

landongrindheim committed:
Remove unused method

landongrindheim committed:
Consider `can_push?` instead of ownership

Ownership is proving to be trickier to keep consistent than I had hoped
it would be. I've been migrating to a permissions-based approach, which
has similar concerns beneath the covers, but is more explicit in its
implementation.

landongrindheim committed:
Remove Rubygem#owned_by?

Ownership has proven to be more complicated than expected, and these
methods were primarily used by our policy code. The logic has been
extracted into GemPermissions, where we've been able to get a bit more
exact.

landongrindheim commented on app/models/gem_permissions.rb:
```diff
@@ -0,0 +1,55 @@
+    end
+  end
+
+  def org_member_in_role?(minimum_role)
+    case minimum_role
+    when :maintainer
+      rubygem.organization.user_is_member?(user)
+    when :admin
+      rubygem.organization.administered_by?(user)
+    when :owner
+      rubygem.organization.owned_by?(user)
+    else
+      false
+    end
+  end
+
+  def user_in_role?(minimum_role)
+    rubygem.ownerships.user_with_minimum_role(user, minimum_role).exists?
+  end
```
I think it makes sense to leave these here _for now_, but I foresee a future where org- (and team-) based permissions are checked in a separate class.

landongrindheim committed:
Test that maintainers cannot invite new members

landongrindheim commented:
> I feel that this Pull Request needs a bit more time to bake before I'm ready to accept it. This is a good step, but I also feel it's not complete either. With Organisation teams coming down the pipeline I feel it would be a good opportunity to spend a bit more time on this?

@colby-swandale I think you're right to be cautious with core permission logic. The organization teams work _will_ add complexity here.

Would it be helpful if I walked through how this is designed to extend? I can show how the permission methods would handle the additional cases we'll need, or we can schedule a quick call to discuss the technical approach.

Happy to adjust the design based on our conversations, or hold off if you'd prefer to finalize the broader requirements first 😄

jenshenny commented on test/system/invitation_test.rb:
```diff
@@ -25,6 +25,17 @@ class InvitationTest < ApplicationSystemTestCase
     assert_text I18n.t("organizations.members.create.member_invited")
   end
 
+  test "maintainers cannot invite users to an organization" do
```
This seems like it can be in another PR too.

jenshenny reviewed (commented):
I like that the gem permissions logic is more centralized with this PR.

landongrindheim commented on app/controllers/api/base_controller.rb:
```diff
@@ -94,8 +94,4 @@ def render_bad_request(error = "bad request")
     render json: { error: error.to_s }, status: :bad_request
   end
-
-  def owner?
```
Yep! I opened https://github.com/rubygems/rubygems.org/pull/6086 for this 😄

landongrindheim committed:
Remove edge case tests

I've lost track of what I was trying to test, and using a non-existing
role results in an error. Removal seems appropriate.

landongrindheim committed:
Merge branch 'master' into landon/push-ownership-logic-to-rubygem-model

landongrindheim committed:
Prevent maintainers (not owners) from transfering
